### PR TITLE
New M3 Demo App for Master Channel

### PR DESCRIPTION
**Pull request by QuncCccccc**
This PR creates a new Material 3 demo app in the experimental directory. The app contains all the updated M3 components in the Flutter master channel.

Screen Shot 2022-08-12 at 12 00 28 PMScreen Shot 2022-08-12 at 12 00 38 PM

## Pre-launch Checklist

- [x] I read the [Flutter Style Guide] _recently_, and have followed its advice.
- [x] I signed the [CLA].
- [x] I read the [Contributors Guide].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.

### CODE REVIEW

File: experimental/material_3_demo/pubspec.yaml
Comment (domesticmouse) [inline experimental/material_3_demo/pubspec.yaml:11]: Any particular reason for `2.17.0-0`? I figure this should really be `2.18.0-0`.

**Review by domesticmouse — APPROVED** [review]

**Comment by domesticmouse** [comment]
Note, this PR is blocked on #1379

**Review by hangyujin — APPROVED** [review]

**Comment by hangyujin** [comment]
Screen Shot 2022-08-15 at 17 58 25

Added textfields too

File: experimental/material_3_demo/pubspec.yaml
Comment (QuncCccccc) [reply experimental/material_3_demo/pubspec.yaml]: Don't have any particular reasons for that version. It's just auto-generated. Thanks for noticing this. Changed!

File: experimental/material_3_demo/lib/main.dart
Comment (HansMuller) [inline experimental/material_3_demo/lib/main.dart:50]: Might be a little more self-explanatory as an enum (modern Dart enums can have properties).

File: experimental/material_3_demo/lib/main.dart
Comment (HansMuller) [inline experimental/material_3_demo/lib/main.dart:147]: two space indent

File: experimental/material_3_demo/lib/main.dart
Comment (HansMuller) [inline experimental/material_3_demo/lib/main.dart:179]: The if clause returns so an else clause isn't needed here.

File: experimental/material_3_demo/lib/main.dart
Comment (HansMuller) [inline experimental/material_3_demo/lib/main.dart:190]: two space indent

File: experimental/material_3_demo/lib/typography_screen.dart
Comment (HansMuller) [inline experimental/material_3_demo/lib/typography_screen.dart:19]: This would be easier to read if all of the TextStyleExample calls were one-liners.

**Review by HansMuller — APPROVED** [review]
This looks good; I've left behind random nit-pick comments just to prove that I read through the code. Other observations:
- Should use single quotes everywhere
- Likewise for trailing commas
- Other than _FooState classes, it probably would be simplest to make everything public. Currently some classes define some private fields, others are all public.

**Comment by domesticmouse** [comment]
Some code style issues will be more obvious once the CI pipeline is integrated. It'll pick up things like single vs double quoted strings.

File: experimental/material_3_demo/lib/component_screen.dart
Comment (QuncCccccc) [reply experimental/material_3_demo/lib/component_screen.dart]: Done!

File: experimental/material_3_demo/lib/component_screen.dart
Comment (QuncCccccc) [reply experimental/material_3_demo/lib/component_screen.dart:357]: Corrected the style.

File: experimental/material_3_demo/lib/component_screen.dart
Comment (QuncCccccc) [reply experimental/material_3_demo/lib/component_screen.dart:61]: This format is generated by the `flutter format` command and it's required if the app will be migrated to non-experimental folder eventually:) Rules are [here](https://github.com/flutter/samples/blob/main/CONTRIBUTING.md#:~:text=Be%20formatted%20with%20flutter%20format)
Will try to reformat it to see if the formatter allows the code to be one line after the CI pipe is ready.

File: experimental/material_3_demo/lib/component_screen.dart
Comment (QuncCccccc) [reply experimental/material_3_demo/lib/component_screen.dart:740]: Same format issue because of the formatter. Will try later.

File: experimental/material_3_demo/lib/component_screen.dart
Comment (QuncCccccc) [reply experimental/material_3_demo/lib/component_screen.dart]: Done. Thanks for the suggestion!

File: experimental/material_3_demo/lib/component_screen.dart
Comment (QuncCccccc) [reply experimental/material_3_demo/lib/component_screen.dart:861]: Also same format issue here.

File: experimental/material_3_demo/lib/component_screen.dart
Comment (QuncCccccc) [reply experimental/material_3_demo/lib/component_screen.dart]: Done. Thanks!

File: experimental/material_3_demo/lib/main.dart
Comment (QuncCccccc) [reply experimental/material_3_demo/lib/main.dart:181]: I think the `else` would be needed when the screen width is greater than 450 pixels:)

**Comment by QuncCccccc** [comment]
> Some code style issues will be more obvious once the CI pipeline is integrated. It'll pick up things like single vs double quoted strings.

Should I give a larger value of the line length to the formatter instead of the default 80, so that some code can fit in one line? like `flutter format -l100 <file>`

**Comment by domesticmouse** [comment]
Please add this project to [`flutter_ci_script_master.sh`](https://github.com/flutter/samples/blob/main/tool/flutter_ci_script_master.sh) to kick in the master CI pipeline.

**Comment by domesticmouse** [comment]
> > Please add this project to [`flutter_ci_script_master.sh`](https://github.com/flutter/samples/blob/main/tool/flutter_ci_script_master.sh) to kick in the master CI pipeline.
> 
> Just added. Please let me know if there's any problems.

Looks like the CI pipeline spat the dummy:

```console
== Testing 'experimental/material_3_demo' on Flutter's master channel ==
~/work/samples/samples/experimental/material_3_demo ~/work/samples/samples
Running "flutter pub get" in material_3_demo...                  2,007ms
Analyzing material_3_demo...

warning - analysis_options.yaml:1:10 - The include file '../analysis_options.yaml' in '/home/runner/work/samples/samples/experimental/material_3_demo/analysis_options.yaml' can't be found when analyzing '/home/runner/work/samples/samples/experimental/material_3_demo'. - include_file_not_found
```

I think you'll find it needs to be `../../analysis_options.yaml`. This is a result of moving the example into the `experimental` directory.

**Comment by domesticmouse** [comment]
[Master CI failure](https://github.com/flutter/samples/runs/7890580024?check_suite_focus=true#step:5:463):

```console
== Testing 'experimental/material_3_demo' on Flutter's master channel ==
~/work/samples/samples/experimental/material_3_demo ~/work/samples/samples
Running "flutter pub get" in material_3_demo...                  1,977ms
Analyzing material_3_demo...

   info - lib/component_screen.dart:260:23 - Prefer const with constant constructors. - prefer_const_constructors
   info - lib/component_screen.dart:285:23 - Prefer const with constant constructors. - prefer_const_constructors
   info - lib/component_screen.dart:312:23 - Prefer const with constant constructors. - prefer_const_constructors
   info - lib/component_screen.dart:825:44 - Avoid annotating types for function expression parameters. - avoid_types_on_closure_parameters

4 issues found.
```

**Comment by domesticmouse** [comment]
LGTM, feel free to land when the other CI steps turn green. Alternatively, I can land this the next time I take a pass through the stack.

File: experimental/material_3_demo/lib/main.dart
Comment (HansMuller) [reply experimental/material_3_demo/lib/main.dart:192]: I meant that this is logically the same:
```dart
if (constraints.maxWidth < narrowScreenWidthThreshold) {
  return Scaffold(
    appBar: createAppBar(),
    body: Row(
      children: <Widget>[
        createScreenFor(ScreenSelected.values[screenIndex], false),
      ],
    ),
    bottomNavigationBar: NavigationBars(
      onSelectItem: handleScreenChanged,
      selectedIndex: screenIndex,
      isExampleBar: false,
    ),
  );
}

return Scaffold(
  appBar: createAppBar(),
  body: SafeArea(
    bottom: false,
    top: false,
    child: Row(
      children: <Widget>[
        Padding(
          padding: const EdgeInsets.symmetric(horizontal: 5),
          child: NavigationRailSection(
            onSelectItem: handleScreenChanged,
            selectedIndex: screenIndex
          ),
        ),
        const VerticalDivider(thickness: 1, width: 1),
        createScreenFor(ScreenSelected.values[screenIndex], true),
      ],
    ),
  ),
);
```

Comment (QuncCccccc) [reply experimental/material_3_demo/lib/main.dart:192]: Oh I got it. You are right. Will fix!!

**Pull request merged**
